### PR TITLE
feat/add vwebhook 4 cluster

### PR DESCRIPTION
**Pull request by steven-zou**

### CODE REVIEW

File: apis/goharbor.io/v1alpha2/harbor_webhook.go [hunk 1]
```diff
@@ -52,18 +52,22 @@ func (h *Harbor) Validate() error {
...
 	if err != nil {
 		allErrs = append(allErrs, field.Invalid(field.NewPath("spec").Child("externalURL"), h.Spec.ExternalURL, err.Error()))
 	}
 
+	if h.Spec.Database == nil {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("spec").Child("database"), h.Spec.Database, "field is required"))
+	}
+
+	if h.Spec.Redis == nil {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("spec").Child("redis"), h.Spec.Redis, "field is required"))
```
Comment (wangcanfengxs) [inline apis/goharbor.io/v1alpha2/harbor_webhook.go:68]: Same comment with above

File: apis/goharbor.io/v1alpha2/harborcluster_types.go [hunk 2]
```diff
@@ -185,6 +204,7 @@ const (
 
 // +kubebuilder:object:root=true
 // +kubebuilder:printcolumn:name="Public URL",type=string,JSONPath=`.spec.externalURL`,description="The public URL to the Harbor application",priority=0
+// +kubebuilder:printcolumn:name="Status", type=string,JSONPath=`.status.status`,description="The overall status of the Harbor cluster",priority=0
```
Comment (wangcanfengxs) [inline apis/goharbor.io/v1alpha2/harborcluster_types.go:207]: .status.status?

File: apis/goharbor.io/v1alpha2/harborcluster_webhook.go [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+		return nil
+	}
+
+	return apierrors.NewInvalid(schema.GroupKind{Group: GroupVersion.Group, Kind: "HarborCluster"}, hc.Name, allErrs)
+}
+
+func (hc *HarborCluster) validateStorage() *field.Error {
+	// Storage
+	// External is not configured
+	if err := hc.Spec.ImageChartStorage.Validate(); err != nil {
```
Comment (wangcanfengxs) [inline apis/goharbor.io/v1alpha2/harborcluster_webhook.go:87]: What will happen if `hc.Spec.ImageChartStorage` is not set and using `hc.Spec.InClusterStorage`

File: apis/goharbor.io/v1alpha2/harborcluster_webhook.go [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+				hc.Spec.InClusterStorage,
+				"conflicts: both storage and in-cluster storage are configured, only one is required to set",
+			)
+		}
+	}
+
+	return nil
+}
+
+func (hc *HarborCluster) validateDatabase() *field.Error {
```
Comment (wangcanfengxs) [inline apis/goharbor.io/v1alpha2/harborcluster_webhook.go:115]: Do we need to validate in cluster database？

File: apis/goharbor.io/v1alpha2/harborcluster_webhook.go [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+			field.NewPath("spec").Child("database", "inClusterDatabase"),
+			hc.Spec.InClusterDatabase,
+			"conflicts: both database or in-cluster database are configured, only one is required to set",
+		)
+	}
+
+	return nil
+}
+
+func (hc *HarborCluster) validateCache() *field.Error {
```
Comment (wangcanfengxs) [inline apis/goharbor.io/v1alpha2/harborcluster_webhook.go:141]: Do we need to validate in cluster cache？

**Review by wangcanfengxs — COMMENTED** [review]
Comments left

File: apis/goharbor.io/v1alpha2/harborcluster_types.go (hunk 2 above)
Comment (steven-zou) [reply apis/goharbor.io/v1alpha2/harborcluster_types.go:208]: yup

File: apis/goharbor.io/v1alpha2/harborcluster_webhook.go (hunk 3 above)
Comment (steven-zou) [reply apis/goharbor.io/v1alpha2/harborcluster_webhook.go:87]: here only check the following two cases:

- both not set
- both are set

File: apis/goharbor.io/v1alpha2/harborcluster_webhook.go (hunk 4 above)
Comment (steven-zou) [reply apis/goharbor.io/v1alpha2/harborcluster_webhook.go:110]: we'll validate all, only one of them configured is valid.

no one is set all both are set are all invalid cases

File: apis/goharbor.io/v1alpha2/harborcluster_webhook.go (hunk 5 above)
Comment (steven-zou) [reply apis/goharbor.io/v1alpha2/harborcluster_webhook.go:132]: see above

File: apis/goharbor.io/v1alpha2/harbor_webhook.go [hunk 6]
```diff
@@ -52,18 +52,22 @@ func (h *Harbor) Validate() error {
 
 	err := h.Spec.ImageChartStorage.Validate()
 	if err != nil {
-		allErrs = append(allErrs, field.Invalid(field.NewPath("spec").Child("persistence").Child("imageChartStorage"), h.Spec.ImageChartStorage, err.Error()))
-	}
-
-	if len(allErrs) == 0 {
-		return nil
+		allErrs = append(allErrs, field.Invalid(field.NewPath("spec").Child("imageChartStorage"), h.Spec.ImageChartStorage, err.Error()))
```
[the remark this answers is not in the document]
Comment (steven-zou) [reply apis/goharbor.io/v1alpha2/harbor_webhook.go:55]: not used agian, so maybe no need to set a var

**Review by Ghostbaby — APPROVED** [review]
lgtm

**Review by chlins — APPROVED** [review]
LGTM

**Pull request merged**